xilinx/src/xlnx_dec_utils.rs

```diff
@@ -0,0 +1,228 @@
+use std::{ffi::CString, os::raw::c_char, str::from_utf8};
+
+use libloading::{Library, Symbol};
+use simple_error::{bail, SimpleError};
+
+use crate::{strcpy_to_arr_i8, sys::*, xrm_precision_1000000_bitmask};
+
+#[derive(Clone, Debug)]
+pub struct XlnxDecBuffer<'a> {
+    pub data: &'a [u8],
+    pub size: usize,
+    pub allocated: usize,
+}
+
+pub struct XlnxDecoderXrmCtx {
+    pub xrm_reserve_id: u64,
+    pub device_id: i32,
+    pub dec_load: i32,
+    pub decode_res_in_use: bool,
+    pub xrm_ctx: xrmContext,
+    pub cu_list_res: xrmCuListResource,
+}
+
+/// Calculates the decoder load uing the xrmU30Dec plugin.
+pub fn xlnx_calc_dec_load(xrm_ctx: xrmContext, xma_dec_props: *mut XmaDecoderProperties) -> Result<i32, SimpleError> {
+    let func_id = 0;
+    let mut plugin_param: xrmPluginFuncParam = Default::default();
+    unsafe {
+        match Library::new("/opt/xilinx/xrm/plugin/libxmaPropsTOjson.so") {
+            Ok(lib) => {
+                match lib
+                    .get::<Symbol<extern "C" fn(props: *mut XmaDecoderProperties, function: *const c_char, json_params: *mut c_char)>>(b"convertXmaPropsToJson")
+                {
+                    Ok(convert_xma_props_to_json) => {
+                        let function = CString::new("DECODER").unwrap();
+                        convert_xma_props_to_json(xma_dec_props, function.as_ptr(), plugin_param.input.as_mut_ptr());
+                    }
+                    Err(e) => bail!("{}", e),
+                }
+            }
+            Err(e) => bail!("{}", e),
+        };
+    }
+
+    let plugin_name = CString::new("xrmU30DecPlugin").unwrap().into_raw();
+
+    unsafe {
+        let ret = xrmExecPluginFunc(xrm_ctx, plugin_name, func_id, &mut plugin_param);
+        if ret != XRM_SUCCESS as i32 {
+            bail!("XRM decoder plugin failed to calculate decoder load. error: {}", ret);
+        }
+    }
+    // parse the load from the output buffer of plugin param.
+    let output_bytes = &plugin_param.output.map(|i| i as u8);
+    let dec_plugin_output = from_utf8(output_bytes).unwrap_or("-1").split(' ').collect::<Vec<&str>>()[0];
```

Comment:
Maybe just use `next` instead of creating a `Vec`:

```suggestion
    let dec_plugin_output = from_utf8(output_bytes).unwrap_or("-1").split(' ').next().expect("split should emit at least one item");
```